device_plugins/nvidia_gpu/nvidia_gpu.go

```diff
@@ -0,0 +1,281 @@
+	_, err = os.Stat(nvidiaUVMToolsDevice)
+	if !os.IsNotExist(err) {
+		ngm.defaultDevices = append(ngm.defaultDevices, nvidiaUVMToolsDevice)
+	}
+
+	if err := ngm.discoverGPUs(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func Register(kubeletEndpoint, pluginEndpoint, resourceName string) error {
+	conn, err := grpc.Dial(kubeletEndpoint, grpc.WithInsecure(),
+		grpc.WithDialer(func(addr string, timeout time.Duration) (net.Conn, error) {
+			return net.DialTimeout("unix", addr, timeout)
+		}))
+	defer conn.Close()
+	if err != nil {
+		return fmt.Errorf("device-plugin: cannot connect to kubelet service: %v", err)
+	}
+	client := pluginapi.NewRegistrationClient(conn)
+	reqt := &pluginapi.RegisterRequest{
+		Version:      pluginapi.Version,
+		Endpoint:     pluginEndpoint,
+		ResourceName: resourceName,
+	}
+
+	_, err = client.Register(context.Background(), reqt)
+	if err != nil {
+		return fmt.Errorf("device-plugin: cannot register to kubelet service: %v", err)
+	}
+	return nil
+}
+
+// Implements DevicePlugin service functions
+func (ngm *nvidiaGPUManager) ListAndWatch(emtpy *pluginapi.Empty, stream pluginapi.DevicePlugin_ListAndWatchServer) error {
+	fmt.Printf("device-plugin: ListAndWatch start\n")
+	changed := true
+	for {
+		for id, dev := range ngm.devices {
+			state := ngm.GetDeviceState(id)
+			if dev.Health != state {
+				changed = true
+				dev.Health = state
+				ngm.devices[id] = dev
+			}
+		}
+		if changed {
+			resp := new(pluginapi.ListAndWatchResponse)
+			for _, dev := range ngm.devices {
+				resp.Devices = append(resp.Devices, &pluginapi.Device{dev.ID, dev.Health})
+			}
+			fmt.Printf("ListAndWatch: send devices %v", resp)
+			if err := stream.Send(resp); err != nil {
+				fmt.Printf("device-plugin: cannot update device states: %v\n", err)
+				ngm.grpcServer.Stop()
+			}
+		}
+		changed = false
+		time.Sleep(5 * time.Second)
+	}
+	return nil
+}
+
+func (ngm *nvidiaGPUManager) Allocate(ctx context.Context, rqt *pluginapi.AllocateRequest) (*pluginapi.AllocateResponse, error) {
+	resp := new(pluginapi.AllocateResponse)
+	for _, id := range rqt.DevicesIDs {
+		dev, ok := ngm.devices[id]
+		if !ok {
+			return nil, fmt.Errorf("Invalid allocation request with non-existing device %s", id)
+		}
+		if dev.Health != pluginapi.Healthy {
+			return nil, fmt.Errorf("Invalid allocation request with unhealthy device %s", id)
+		}
+		devRuntime := new(pluginapi.DeviceRuntimeSpec)
+		devRuntime.Devices = append(devRuntime.Devices, &pluginapi.DeviceSpec{
+			HostPath:      "/dev/" + id,
+			ContainerPath: "/dev/" + id,
+			Permissions:   "mrw",
+		})
+		for _, d := range ngm.defaultDevices {
+			devRuntime.Devices = append(devRuntime.Devices, &pluginapi.DeviceSpec{
+				HostPath:      d,
+				ContainerPath: d,
+				Permissions:   "mrw",
+			})
+		}
+		devRuntime.Mounts = append(devRuntime.Mounts, &pluginapi.Mount{
+			ContainerPath: "/usr/local/nvidia/lib64",
+			HostPath:  "/home/kubernetes/bin/nvidia/lib",
```

Comment:
Can we use const to represent base paths? Make the base paths configurable to make the plugin work outside of COS.

Reply:
done.